### PR TITLE
feat: System wide typed pub sub topic registry

### PR DESCRIPTION
**Pull request by johanandren**
On top of #32261 

References #31053

### CODE REVIEW

**Review by patriknw — COMMENTED** [review]
looking good, docs should also be updated

**Comment by johanandren** [comment]
Something I've thought about during the weekend:

There is a race between TTL shutdown and the janitor actor cleaning up, we could potentially do better by having an I'm-going-to-stop callback to the Topic that it calls just before shutting down and then have a retry spin if the name is still occupied but not in the registry where the actor will soon shut down.

Not entirely sure it is worth it though.

**Comment by jroper** [comment]
One race that I think would be good to address is the first publish race. I think, when you first publish to a topic that is inactive, there's a race between the message you send, and the first `Listing` sent by the receptionist, and I'm pretty sure the message you send will nearly always win, so that means that first message sent to a topic that is inactive on the node, but has subscribers elsewhere in the cluster, will never almost never make it to the subscribers. If we could stash publish messages until the first listing is received, I think that would solve it.

**Comment by johanandren** [comment]
In the typed group topic, so I think we can do that, but let's treat it as a separate issue. I created https://github.com/akka/akka/issues/32267 for that.

File: akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    if (ttlInRegistry != ttl)
+      log.warn(
+        "Asked for topic [{}] with TTL [{}] but already existing topic has a different TTL [{}]",
+        name,
+        ttl.map(_.pretty).getOrElse("none"),
+        ttlInRegistry.map(_.pretty).getOrElse("none"))
+    actorRefInRegistry.narrow
+  }
+
+  private def spawnTopic[T](name: String, classTag: ClassTag[T], ttl: Option[FiniteDuration]): ActorRef[Any] = {
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala:134]: maybe we can make it less untyped? here is a suggestion that compiles a0aa9a9843

**Review by patriknw — COMMENTED** [review]
looking good

File: akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala]: Included

File: akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      classTag: ClassTag[T]): ActorRef[Topic.Command[T]] = {
+    val messageClass = classTag.runtimeClass
+    val (classInRegistry, ttlInRegistry, actorRefInRegistry) =
+      registry.computeIfAbsent(
+        name,
+        name => (messageClass.asInstanceOf[Class[Any]], ttl, spawnTopic(name, classTag, ttl)))
+    if (classInRegistry != messageClass)
+      throw new IllegalArgumentException(
+        s"Trying to start topic [$name] with command class [$messageClass], but it is was already started with a " +
+        s"different command class [$classInRegistry]. Use unique topic names for all your topics.")
```
Comment (octonato) [inline akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala:123]: Isn't this exception happening a little too late? 

If I create a new topic using a different `name` but for an already registered `messageClass`, the function that we pass to `computeIfAbsent` will be executed, the new actor will be spawned, but then we will throw an exception.

**Review by octonato — COMMENTED** [review]
Looking good, but I think I spotted an issue.

**Review by patriknw — COMMENTED** [review]
update reference docs?

File: akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala:123]: Different topic name using the with same type of message is fine. Topic name needs to be globally unique and cannot be re-used for different protocols. 

This is here to catch accidental use of the same name for different types of protocol. If there already is an entry for a name `computeIfAbsent` doesn't change anything but returns the existing entry, if that is for a different type of messages failing here makes sure users notice.

File: akka-docs/src/main/paradox/typed/distributed-pub-sub.md [hunk 3]
```diff
@@ -26,22 +26,25 @@ Additionally, add the dependency as below.
...
 
-Distributed publish subscribe is achieved by representing each pub sub topic with an actor, @apidoc[akka.actor.typed.pubsub.Topic](akka.actor.typed.pubsub.Topic$). 
+The topic actor needs to run on each node where subscribers will live or that wants to publish messages to the topic. 
 
-The topic actor needs to run on each node where subscribers will live or that wants to publish messages to the topic.
- 
-The identity of the topic is a tuple of the type of messages that can be published and a string topic name but it is recommended
-to not define multiple topics with different types and the same topic name.
+Topics can be looked up in the @apidoc[PubSub](akka.actor.typed.pubsub.PubSub$) registry, this way the same topic will be represented by the same actor for a whole
+actor system. If the topic has not yet been started.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/distributed-pub-sub.md:36]: Is that last sentence incomplete?

**Review by patriknw — APPROVED** [review]
LGTM after a minor doc comment

File: akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala (hunk 2 above)
Comment (octonato) [reply akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala:123]: Correct, I don't know how I missed the fact that `classInRegistry` will always be for either a new name or an already registered name and not the other way around.

Comment (octonato) [inline akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala:123]:
```suggestion
        s"Trying to start topic [$name] with command class [$messageClass], but it was already started with a " +
        s"different command class [$classInRegistry]. Use unique topic names for all your topics.")
```

File: akka-docs/src/main/paradox/typed/distributed-pub-sub.md (hunk 3 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/distributed-pub-sub.md]: Yea, halfway through a thought. Fixed now.

**Pull request merged**